zebROS_ws/src/controllers_2023/src/four_bar_controller.cpp

```diff
@@ -0,0 +1,461 @@
+
+namespace four_bar_controller_2023
+{
+
+class FourBarCommand_2023
+{
+    public:
+        FourBarCommand_2023()
+        {
+            angle_ = 0;
+        }
+        FourBarCommand_2023(double angle)
+        {
+            angle_ = angle;
+        }
+        double GetAngle() const
+        {
+            return angle_;
+        }
+
+
+    private:
+        double angle_; // radians.
+        // lower limit = 0 (all the way to the left), upper limit = pi (all the way to the right)
+};
+//this is the actual controller, so it stores all of the  update() functions and the actual handle from the joint interface
+//if it was only one type, controller_interface::Controller<TalonCommandInterface> here
+class FourBarController_2023 : public controller_interface::MultiInterfaceController<hardware_interface::TalonCommandInterface>
+
+{
+    public:
+        FourBarController_2023()
+        {
+        }
+
+        //should this be hardware_interface::TalonCommandInterface instead? What's the reason to import RobotHW then get CommandInterface from that instead of just importing TalonCommandIface?
+        //answer to my question: the TalonCommandInterface is passed in if it's not a multiInterfaceController, and just one kind of joint is made!
+        virtual bool init(hardware_interface::RobotHW *hw,
+                          ros::NodeHandle             &root_nh,
+                          ros::NodeHandle             &controller_nh) override;
+        virtual void starting(const ros::Time &time) override;
+        virtual void update(const ros::Time &time, const ros::Duration &period) override;
+        virtual void stopping(const ros::Time &time) override;
+
+        bool cmdService(controllers_2023_msgs::FourBarSrv::Request &req,
+                        controllers_2023_msgs::FourBarSrv::Response &res);
+
+        double angleFromX(double x, bool below) {
+            // motor reads clockwise as positive, but angles are counterclockwise.
+            // so, this angle needs to be negative.
+            double angle = acos((min_extension_-intake_length_-parallel_bar_length_)/diagonal_bar_length_) - acos((x-intake_length_-parallel_bar_length_)/diagonal_bar_length_);
+            return below ? acos((min_extension_-intake_length_-parallel_bar_length_)/diagonal_bar_length_) + acos((x-intake_length_-parallel_bar_length_)/diagonal_bar_length_) : angle;
+        }
+
+
+    private:
+        ros::Time last_time_down_;
+        talon_controllers::TalonControllerInterface four_bar_joint_; //interface for the talon joint
+
+        realtime_tools::RealtimeBuffer<FourBarCommand_2023> position_command_; //this is the buffer for percent output commands to be published
+        ros::ServiceServer four_bar_service_; //service for receiving commands
+
+        bool zeroed_;
+        bool last_zeroed_;
+        double last_angle_;
+        //double last_setpoint_;
+        hardware_interface::TalonMode last_mode_;
+
+        double max_extension_;
```

Comment:
Do you want to make this dynamic reconfigurable, to match everything else?

Reply:
sure, good idea